### PR TITLE
Add Django authentication script

### PR DESCRIPTION
**Pull request by countermeasure**
An authentication script for Django apps, as mentioned a few hours ago in the `zaproxy-users` group.

### CODE REVIEW

File: authentication/DjangoAuthentication.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
+/*
+ * A script to provide authentication for Django apps.
+ *
+ * First it makes a GET request and obtains the csrfmiddlewaretoken from the response body.
+ *
+ * Then it makes a POST request with a body which contains username, password and csrfmiddlewaretoken.
+ *
+ * A successful login will result in a 302 redirect. If this happens, a GET request is made to the redirect URL.
+ *
+ * Every request made by this script is logged seperately to the History tab.
```
Comment (thc202) [inline authentication/DjangoAuthentication.js]: There's a typo in separately.

Comment (countermeasure) [reply authentication/DjangoAuthentication.js]: I think I'm destined to always misspell that word!

What's the best way for me to fix it? Should I amend the commit and force push, or make the correction in a new commit?

Comment (kingthorin) [reply authentication/DjangoAuthentication.js]: Amend and force is best, to maintain an atomic commit.
https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#what-we-zap-team-expect-from-you

Comment (countermeasure) [reply authentication/DjangoAuthentication.js]: For sure. Thanks for drawing my attention to the contributing document.

I just force pushed then.

**Comment by thc202** [comment]
Thanks!

**Comment by countermeasure** [comment]
Thanks for merging @kingthorin and @thc202!

**Comment by kingthorin** [comment]
Thanks for the contribution!

**Comment by huornlmj** [comment]
@countermeasure - can you explain how one might use this?

**Comment by thc202** [comment]
The form-based authentication will be better than using the script, it now supports anti-CSRF tokens (and Django's one is added by default).

I'd suggest reading this blog post though: https://www.zaproxy.org/blog/2023-05-02-authentication-auto-detection/

**Comment by huornlmj** [comment]
I read the blog and it appears to be more for automation then manual testing.

**Comment by countermeasure** [comment]
Hi @huornlmj 👋🏼 
I'm no expert on this, but I can tell you what I know.
I just use it in automated tests, although I think I remember using it for manual testing at one stage a while back.
Here's an example of my current automation config, which gets the job done for me:
```
---
env:
  contexts :
    - name: "authenticated"
      urls:
        - "https://example.com"
      includePaths:
        - "https://example.com.*"
      authentication:
        method: "script"
        parameters:
          script: "/path/to/.ZAP/community-scripts/authentication/DjangoAuthentication.js"
          scriptEngine: "Oracle Nashorn"
          Target URL: "https://example.com/accounts/login/"
          Username field: "username"
          Password field: "password"
        verification:
          method: "response"
          loggedInRegex: "Log Out"
          loggedOutRegex: "Log In"
      sessionManagement:
        method: "cookie"
      users:
        - name: "User"
          credentials:
            Username: "my_username"
            Password: "my_password"
  parameters:
    progressToStdout: true
jobs:
  - type: spider
    parameters:
      maxDuration: 5
      user: "User"
      url: "https://example.com"
  - type: activeScan
    parameters:
      maxScanDurationInMins: 5
      user: "User"
    policyDefinition:
      defaultStrength: Medium
      defaultThreshold: Medium
  - type: report
    parameters:
      template: modern
      reportDir: "/path/to/test_reports/zap"
      reportFile: "authenticated_scan_report.html"
      reportTitle: "Authenticated Scan"
```
More info on how configuration works is at https://www.zaproxy.org/docs/desktop/addons/automation-framework/environment/.

**Pull request merged**